core/rest-client/src/http_client.rs

```diff
@@ -0,0 +1,425 @@
+	response::{Headers, Response},
+	uri::Uri,
+};
+use log::*;
+use std::{
+	collections::HashMap,
+	str::FromStr,
+	string::{String, ToString},
+	time::Duration,
+	vec::Vec,
+};
+use url::Url;
+
+pub type EncodedBody = Vec<u8>;
+
+/// Simple trait to send HTTP request, based on `http_req`.
+///
+/// Automatically upgrades to TLS in case the base URL contains 'https'
+pub trait SendHttpRequest {
+	fn send_request<U, T>(
+		&self,
+		base_url: Url,
+		method: Method,
+		params: U,
+		query: Option<&Query<'_>>,
+		maybe_body: Option<String>,
+	) -> Result<(Response, EncodedBody), Error>
+	where
+		T: RestPath<U>;
+}
+
+/// HTTP client implementation
+///
+/// wrapper for the `http_req` library that adds the necessary headers and body to a request
+pub struct HttpClient {
+	send_null_body: bool,
+	timeout: Option<Duration>,
+	headers: Headers,
+	authorization: Option<String>,
+}
+
+impl Default for HttpClient {
+	fn default() -> Self {
+		HttpClient {
+			send_null_body: true,
+			timeout: None,
+			headers: Headers::new(),
+			authorization: None,
+		}
+	}
+}
+
+impl HttpClient {
+	pub fn new(
+		send_null_body: bool,
+		timeout: Option<Duration>,
+		headers: Option<Headers>,
+		authorization: Option<String>,
+	) -> Self {
+		HttpClient {
+			send_null_body,
+			timeout,
+			headers: headers.unwrap_or_else(Headers::new),
+			authorization,
+		}
+	}
+
+	/// Set credentials for HTTP Basic authentication.
+	pub fn set_auth(&mut self, user: &str, pass: &str) {
+		let mut s: String = user.to_string();
+		s.push(':');
+		s.push_str(pass);
+		self.authorization = Some(format!("Basic {}", base64::encode(&s)));
+	}
+
+	/// Set HTTP header from string name and value.
+	///
+	/// The header is added to all subsequent GET and POST requests
+	/// unless the headers are cleared with `clear_headers()` call.
+	pub fn set_header(&mut self, name: &'static str, value: &str) -> Result<(), Error> {
+		let header_name = HeaderName::from_str(name).map_err(|_| Error::InvalidValue)?;
+		let value = HeaderValue::from_str(value).map_err(|_| Error::InvalidValue)?;
+
+		add_to_headers(&mut self.headers, header_name, value);
+		Ok(())
+	}
+
+	/// Clear all previously set headers
+	pub fn clear_headers(&mut self) {
+		self.headers = Headers::new();
+	}
+}
+
+impl SendHttpRequest for HttpClient {
+	fn send_request<U, T>(
+		&self,
+		base_url: Url,
+		method: Method,
+		params: U,
+		query: Option<&Query<'_>>,
+		maybe_body: Option<String>,
+	) -> Result<(Response, EncodedBody), Error>
+	where
+		T: RestPath<U>,
+	{
+		let url = join_url(base_url, T::get_path(params)?.as_str(), query)?;
+		let uri = Uri::from_str(url.as_str()).map_err(Error::HttpReqError)?;
+
+		trace!("uri: {:?}", uri);
+
+		let mut request = Request::new(&uri);
+		request.method(method);
+
+		let mut request_headers = Headers::default_http(&uri);
+
+		// have to make this copy because the request only takes the body as reference, not owned
+		let body_encoded = maybe_body.clone().unwrap_or_default().encode();
```

Comment:
I don't really understand. From reading the code, I believe this step is unnecessary.

Reply:
Which step do you mean? The clone? Or the variable in general? The issue here is, that further below I'm doing some logic with `maybe_body` that requires moves. But I need an instance that lives long enough so it can be added to the request, inside that block of logic. It's really not nice, that's true - but I don't see another way, since `Request` of `http_req` does not take ownership of anything you pass in, which makes it quite messy imo.

Reply:
Ah ok, I did somehow not get that.

I see the problem now, the extracted `body` is dropped after the if statement. I think you should be able to solve that with an:

```
if let Some(body) = maybe_body.as_ref() 
```

or something the like

Reply:
Now that we're discussing this issue, I think this is not correct anyway 🙈  I should not be calling `encode()` but rather `as_bytes()`, since the http client does no expect the string encoded, but just as plain utf-8 (as can be seen when receiving the body in the response). I need to add a test that does a `POST` request with a body.

However, I think I still need the clone though, because I move the body with `maybe_body.unwrap_or_default()` (without the `clone()`) and it would no longer be accessible afterwards. I totally agree that it's not ideal, I would like to have a cleaner way, but I don't see it at the moment.

Reply:
This is what I was suggesting. I tested it locally and it compiles. 😄 

```
		if let Some(body) = maybe_body.as_ref() {
			if self.send_null_body || body != "null" {
				let len = HeaderValue::from_str(&body.len().to_string())
					.map_err(|_| Error::RequestError)?;

				add_to_headers(&mut request_headers, CONTENT_LENGTH, len);
				add_to_headers(
					&mut request_headers,
					CONTENT_TYPE,
					HeaderValue::from_str("application/json").unwrap(),
				);

				trace!("set request body: {}", body);
				request.body(body.as_bytes()); // takes body non-owned (!)
			}
```